lib/auth/oidc.go

```diff
@@ -476,6 +477,111 @@ func claimsFromUserInfo(oidcClient *oidc.Client, issuerURL string, accessToken s
 	return claims, nil
 }
 
+func (a *AuthServer) claimsFromGSuite(oidcClient *oidc.Client, issuerURL string, userEmail string, accessToken string) (jose.Claims, error) {
+	err := isHTTPS(issuerURL)
+	if err != nil {
+		return nil, trace.Wrap(err)
+	}
+
+	oac, err := oidcClient.OAuthClient()
+	if err != nil {
+		return nil, trace.Wrap(err)
+	}
+	hc := oac.HttpClient()
+
+	u, err := url.Parse(teleport.GSuiteGroupsEndpoint)
+	if err != nil {
+		return nil, trace.Wrap(err)
+	}
+
+	fetchGroups := func(pageToken string) (*gsuiteGroups, error) {
+		q := u.Query()
+		q.Set("userKey", userEmail)
+		if pageToken != "" {
+			q.Set("pageToken", pageToken)
+		}
+		u.RawQuery = q.Encode()
+		endpoint := u.String()
+
+		log.Debugf("Fetching OIDC claims from GSuite groups endpoint: %q.", endpoint)
+
+		req, err := http.NewRequest("GET", endpoint, nil)
+		if err != nil {
+			return nil, trace.Wrap(err)
+		}
+		req.Header.Set("Authorization", fmt.Sprintf("Bearer %s", accessToken))
+
+		resp, err := hc.Do(req)
+		if err != nil {
+			return nil, trace.Wrap(err)
+		}
+		defer resp.Body.Close()
+
+		bytes, err := ioutil.ReadAll(resp.Body)
+		if err != nil {
+			return nil, trace.Wrap(err)
+		}
+
+		if resp.StatusCode < 200 || resp.StatusCode > 299 {
+			return nil, trace.AccessDenied("bad status code: %v %v", resp.StatusCode, string(bytes))
+		}
+
+		var response gsuiteGroups
+		if err := json.Unmarshal(bytes, &response); err != nil {
+			return nil, trace.BadParameter("failed to parse response: %v", err)
+		}
+		return &response, nil
+	}
+
+	count := 0
+	var groups []string
+	var nextPageToken string
+collect:
```

Comment:
I'd break this out into it's own function, seems like a lot packed into a single function.